### DISCUSSION TITLE
How Do You Clip a MultiBlock?

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [pyvista/pyvista#3618] [Q&A]
I want to slice all elements of a `MultiBlock` with `add_mesh_clip_plane`. To do so, I `combine()` my block into an `UnstructuredGrid`, but the colors I set on the individual elements are lost in the process. How can I clip a `MultiBlock` such that colors of the elements are preserved?

```python
import pyvista as pv

p = pv.Plotter()

blocks = pv.MultiBlock()

blocks.append(pv.Sphere())
blocks.append(pv.Cube(center=(0, 0, -1)))

actor, mapper = p.add_composite(blocks)
mapper.block_attr[0].color = 'red'
mapper.block_attr[1].color = 'blue'

combined = blocks.combine()
	
widget = p.add_mesh_clip_plane(combined)
p.remove_actor(actor)
	
p.show()
```

![image](https://user-images.githubusercontent.com/59346180/202591736-26ac5a29-ad19-4028-bc19-65fbbc1fef2d.png)

**Comment by adam-grant-hendry** [comment]
An inelegant workaround is to use a `KDTree` with a custom color map. This slows down processing and only works to transfer colors. Other scalars are lost.

```python
import numpy as np
import pyvista as pv

from matplotlib.colors import ListedColormap

red = np.array([1.0, 0.0, 0.0, 1.0])
blue = np.array([12 / 256, 238 / 256, 246 / 256, 1.0])

mapping = np.linspace(0, 1, 256)
new_colors = np.empty((256, 4))

new_colors[mapping == 0] = red
new_colors[mapping == 1] = blue

cmap = ListedColormap(new_colors)

p = pv.Plotter()

sphere = pv.Sphere()
cube = pv.Cube(center=(0, 0, -1))

block = pv.MultiBlock()

block.append(sphere)
block.append(cube)

actor, mapper = p.add_composite(block)

combined = block.combine()

tree = KDTree(combined.points.astype(np.double))
_, inds = tree.query(sphere.points)
	
combined.point_data['Scalars'] = np.zeros(combined.points.shape)
combined.point_data['Scalars'][inds] = 1
	
widget = p.add_mesh_clip_plane(combined, cmap=cmap)
p.remove_actor(actor)

p.show()
```

![image](https://user-images.githubusercontent.com/59346180/202596047-9cb7b9aa-bad6-488a-9b00-1ae733a3fc25.png)

**Reply by adam-grant-hendry** [reply]
Actually, since we are just masking, we can make the color space binary:

```python
red = np.array([1.0, 0.0, 0.0, 1.0])
blue = np.array([0.0, 0.0, 1.0, 1.0])

mapping = np.linspace(0, 1, 2)
new_colors = np.empty((2, 4))

...

sbar_args = dict(n_labels=2,)

widget = p.add_mesh_clip_plane(combined, cmap=cmap, scalar_bar_args=sbar_args)
```

![image](https://user-images.githubusercontent.com/59346180/202598843-01fa34a6-ca57-43c1-a2b4-f42249d7ebe7.png)

**Reply by adam-grant-hendry** [reply]
@banesullivan Is there a way we can use `MultiBlock` datasets with `add_mesh_clip_plane()`?

**Reply by banesullivan** [reply]
This appears broken at this time. I'll see if I can add a fix while doing some refactoring to these methods in #3318
